wallet/update.go

```diff
@@ -29,153 +29,141 @@ type (
 		RemoveSiafundElements([]types.SiafundElement, types.ChainIndex) error
 
 		AddressRelevant(types.Address) (bool, error)
-	}
-
-	// An ApplyTx atomically applies a set of updates to a store.
-	ApplyTx interface {
-		UpdateTx
 
 		ApplyMatureSiacoinBalance(types.ChainIndex) error
 		AddEvents([]Event) error
-	}
-
-	// RevertTx atomically reverts an update from a store.
-	RevertTx interface {
-		UpdateTx
 
 		RevertMatureSiacoinBalance(types.ChainIndex) error
 		RevertEvents(index types.ChainIndex) error
 	}
 )
 
-// ApplyChainUpdates atomically applies a set of chain updates to a store
-func ApplyChainUpdates(tx ApplyTx, updates []chain.ApplyUpdate) error {
-	for _, cau := range updates {
-		// update the immature balance of each relevant address
-		if err := tx.ApplyMatureSiacoinBalance(cau.State.Index); err != nil {
-			return fmt.Errorf("failed to get matured siacoin elements: %w", err)
-		}
+// applyChainUpdate atomically applies a chain update to a store
+func applyChainUpdate(tx UpdateTx, cau chain.ApplyUpdate) error {
+	// update the immature balance of each relevant address
+	if err := tx.ApplyMatureSiacoinBalance(cau.State.Index); err != nil {
+		return fmt.Errorf("failed to get matured siacoin elements: %w", err)
+	}
 
-		// determine which siacoin and siafund elements are ephemeral
-		//
-		// note: I thought we could use LeafIndex == EphemeralLeafIndex, but
-		// it seems to be set before the subscriber is called.
-		created := make(map[types.Hash256]bool)
-		ephemeral := make(map[types.Hash256]bool)
-		for _, txn := range cau.Block.Transactions {
-			for i := range txn.SiacoinOutputs {
-				created[types.Hash256(txn.SiacoinOutputID(i))] = true
-			}
-			for _, input := range txn.SiacoinInputs {
-				ephemeral[types.Hash256(input.ParentID)] = created[types.Hash256(input.ParentID)]
-			}
-			for i := range txn.SiafundOutputs {
-				created[types.Hash256(txn.SiafundOutputID(i))] = true
-			}
-			for _, input := range txn.SiafundInputs {
-				ephemeral[types.Hash256(input.ParentID)] = created[types.Hash256(input.ParentID)]
-			}
+	// determine which siacoin and siafund elements are ephemeral
+	//
+	// note: I thought we could use LeafIndex == EphemeralLeafIndex, but
+	// it seems to be set before the subscriber is called.
```

Comment:
yeah, even ephemeral outputs are added to the Merkle tree, so they are assigned a proper leaf index.

It is kind of annoying that you have to manually check them like this Maybe we can add an `ephemeral` bool to `ForEachSiacoinOutput` or something.